### DISCUSSION TITLE
Network-based optimal facility location modeling

### DISCUSSION THREAD

**Discussion by gelazari** [pysal/spaghetti#563] [General] [answered]
Hi all,

I hope this email finds you well. I recently came across your post with respect to Demostrating network-based optimal facility location modeling. I am a PhD student at University College London and I want to use this code as a preprocessing step for a deep learning pipeline I am building. Naturally, this is not my area of expertise. I was wondering if the same code could be used with lat-lon locations and a city map instead of the "shape" data. I have a list of clients, their demand and their (lat, lon) coordinates, and I would like to use spaghetti to find the optimal number and location of possible factories to be deployed in the area.

Thank you in advance for your time and consideration. Looking forward to your response.

**Comment by jGaboardi** [comment, marked as answer]
Thanks for starting this discussion @gelazari.

1. To start off, (and to be clear) no mathematical optimization actually happens in `spaghetti` (see also [this comment](https://github.com/pysal/spaghetti/discussions/554#discussioncomment-135541)). The optimal facility location modeling is simply a tutorial/demo for integrating the output of `spaghetti` as input for optimization problems utilizing other Python libraries/wrappers. That being said, you are free to use any code in the `spaghetti` docs as you like (we are under a [BSD 3-Clause license](https://github.com/pysal/spaghetti/blob/main/LICENSE.txt)), but please cite `spaghetti` appropriately. In the (near?) future, the optimization code will actually find a home in another PySAL project ([`spopt`](https://github.com/pysal/spopt)), so check there from time to time.
2. Any type of coordinates will work, so long as all datasets are of the same coordinate systems & projections (e.g. point locations and streets in the same system). I am not sure what you mean by city map vs. shape data. The input data to create a `spaghetti.Network` object can either be a line-based shapefile, `geopandas.GeoDataFrame`, or collection of `libpysal` geometries, as described [in the docs](https://pysal.org/spaghetti/generated/spaghetti.Network.html#spaghetti.Network). And the input data to create a `spaghetti.PointPattern` object can either be a point-based shapefile, `geopandas.GeoDataFrame`, or collection of `libpysal` geometries, as described [in the docs](https://pysal.org/spaghetti/generated/spaghetti.PointPattern.html#spaghetti.PointPattern). The inputs for the optimization problems are also described [in the tutorials](https://pysal.org/spaghetti/notebooks/facility-location.html#1.3.-Define-a-class-and-functions-for-solving-the-models-and-analyzing-the-results).
3. Since you are dealing with client demand, either the [*p*-median problem](https://pysal.org/spaghetti/references.html#hakimi1964) ([info](https://pysal.org/spaghetti/notebooks/facility-location.html#3.2-p-median-Problem)) or [*MCLP*](https://pysal.org/spaghetti/references.html#church1974) ([info](https://pysal.org/spaghetti/notebooks/facility-location.html#3.4-Maximal-Covering-Location-Problem)) would be appropriate depending on your desired objective (see [this comment](https://github.com/pysal/spaghetti/discussions/553#discussioncomment-135533) for more info).

Hope this helps! If it (or other discussion threads) do help, please mark as "answered."

**Reply by gelazari** [reply]
Hi James, thanks for your speedy and precise reply.

1. From what I saw, the optimization is done using Google's ortools --correct me if I'm wrong. Naturally, 'spaghetti' will be indeed cited if I manage to use the code for my preprocessing purposes.

2. If it is possible, I would like a more in-depth explanation on how I can use a GeoDataFrame (what should be changed/added/removed in the pipeline). I am converting my pandas.DataFrame which contains N clients, N lat, N lon and N demands into a GeoDataFrame. Would that be enough?

3. Thanks for pointing them out, I've already familiarised myself with the papers/problems. From what I understand, I can use the example code to tackle my issue, right? I'd be more than happy if you would help me with that, since it is completely out of my expertise.
4. 
Thank you for your time and consideration. Looking forward to your response.

**Reply by jGaboardi** [reply]
1. Yes, `ortools` is used in the Facility Location tutorial.
2. This depends on where you are getting your data and what format it is in. [Geopandas](https://geopandas.readthedocs.io/en/latest/) has a ton of information on [the creation and manipulation of `GeoDataFrame` objects](https://geopandas.readthedocs.io/en/latest/getting_started.html) with tutorials. Specifically, the [`geopandas.points_from_xy()`](https://geopandas.org/gallery/create_geopandas_from_pandas.html#from-longitudes-and-latitudes) function would be particularly helpful.
3. Yes, the example code should get you a good way there. Give the workflow a try with your data, etc. and see if you can figure it out. Once you have tried and if you are having trouble, try StackExchange ([stackoverflow](https://stackoverflow.com), [GIS](https://gis.stackexchange.com), etc.) then come back here as a last resort with the code you attempted the problems with.
